google-cloud-firestore/src/main/java/com/google/cloud/firestore/CollectionReference.java

```diff
@@ -129,53 +128,59 @@ public DocumentReference document(@Nonnull String childPath) {
    */
   @Nonnull
   public Iterable<DocumentReference> listDocuments() {
-    ListDocumentsRequest.Builder request = ListDocumentsRequest.newBuilder();
-    request.setParent(options.getParentPath().toString());
-    request.setCollectionId(options.getCollectionId());
-    request.setMask(DocumentMask.getDefaultInstance());
-    request.setShowMissing(true);
-
-    final ListDocumentsPagedResponse response;
-    final TraceUtil traceUtil = TraceUtil.getInstance();
-    Span span = traceUtil.startSpan(TraceUtil.SPAN_NAME_LISTDOCUMENTS);
-    try (Scope scope = traceUtil.getTracer().withSpan(span)) {
+    TraceUtil.Span span =
+        rpcContext
+            .getFirestore()
+            .getOptions()
+            .getTraceUtil()
+            .startSpan(TraceUtil.SPAN_NAME_COL_REF_LIST_DOCUMENTS);
+    try (Scope ignored = span.makeCurrent()) {
+      ListDocumentsRequest.Builder request = ListDocumentsRequest.newBuilder();
+      request.setParent(options.getParentPath().toString());
+      request.setCollectionId(options.getCollectionId());
+      request.setMask(DocumentMask.getDefaultInstance());
+      request.setShowMissing(true);
+      final ListDocumentsPagedResponse response;
       FirestoreRpc client = rpcContext.getClient();
       UnaryCallable<ListDocumentsRequest, ListDocumentsPagedResponse> callable =
           client.listDocumentsPagedCallable();
       ListDocumentsRequest build = request.build();
       ApiFuture<ListDocumentsPagedResponse> future = rpcContext.sendRequest(build, callable);
       response = ApiExceptions.callAndTranslateApiException(future);
+      Iterable<DocumentReference> result =
+          new Iterable<DocumentReference>() {
+            @Override
+            @Nonnull
+            public Iterator<DocumentReference> iterator() {
+              final Iterator<Document> iterator = response.iterateAll().iterator();
+              return new Iterator<DocumentReference>() {
+                @Override
+                public boolean hasNext() {
+                  return iterator.hasNext();
+                }
+
+                @Override
+                public DocumentReference next() {
+                  ResourcePath path = ResourcePath.create(iterator.next().getName());
+                  return document(path.getId());
+                }
+
+                @Override
+                public void remove() {
+                  throw new UnsupportedOperationException("remove");
+                }
+              };
+            }
+          };
+      span.end();
+      return result;
     } catch (ApiException exception) {
-      span.setStatus(Status.UNKNOWN.withDescription(exception.getMessage()));
+      span.end(exception);
       throw FirestoreException.forApiException(exception);
-    } finally {
```

Comment:
Why remove the `finally`? (Though the new implementation looks more correct since it doesn't call `end` multiple times when no exception)